src/uct/ib/rdmacm/rdmacm_cm_ep.c

```diff
@@ -192,27 +293,66 @@ static ucs_status_t uct_rdmacm_cm_create_dummy_qp(struct rdma_cm_id *id,
     return UCS_OK;
 }
 
-static ucs_status_t
-uct_rdamcm_cm_ep_set_qp_num(struct rdma_conn_param *conn_param,
-                            uct_rdmacm_cm_ep_t *cep)
+static ucs_status_t uct_rdamcm_cm_ep_create_qpn(uct_rdmacm_cm_ep_t *cep,
+                                                uint32_t* qpn)
 {
-    struct ibv_cq *cq;
+    struct uct_rdmacm_cm_qp_context *ctx;
     ucs_status_t status;
 
-    status = uct_rdmacm_cm_get_cq(uct_rdmacm_cm_ep_get_cm(cep), cep->id->verbs,
-                                  &cq);
+    status = uct_rdmacm_cm_get_qp_context(uct_rdmacm_cm_ep_get_cm(cep),
+                                          cep->id->verbs, &ctx);
     if (status != UCS_OK) {
-        return status;
+        ucs_error("failed to get qp context");
+        goto out;
     }
 
-    /* create a dummy qp in order to get a unique qp_num to provide to librdmacm */
-    status = uct_rdmacm_cm_create_dummy_qp(cep->id, cq, &cep->qp);
-    if (status != UCS_OK) {
-        return status;
+    if (ctx->reserved_qpn_support) {
+        status = uct_rdamcm_cm_ep_create_reserved_qpn(cep, ctx);
+        if (status == UCS_OK) {
+            cep->flags |= UCT_RDMACM_CM_EP_QPN_CREATED;
+            *qpn = cep->qpn;
```

Comment:
seems like it should be part of a caller `uct_rdamcm_cm_ep_set_qp_num` as well as line 321

Reply:
@evgeny-leksikov Fixed.